pkg/apis/resourcedefinition/helper.go

```diff
@@ -0,0 +1,506 @@
+		if !jp.IsArray() || jp.Raw == "[]" {
+			return
+		}
+
+		// Default the element of array with the first available item.
+		var ri gjson.Result
+		for i, rs := 0, jp.Array(); i < len(rs) && !ri.Exists(); i++ {
+			ri = rs[i]
+		}
+
+		if !ri.Exists() {
+			return
+		}
+
+		defaultVariablesSchemaRef(vsr.Value.Items, []byte(ri.Raw))
+	}
+}
+
+// alignSchemas aligns the schemas,
+// which is performs on the first-level properties of the variables schemas and outputs schemas.
+func alignSchemas(scs []openapi3.Schemas) openapi3.Schemas {
+	var (
+		ret = openapi3.Schemas{}
+		nb  = map[string]any{}
+	)
+
+	for _, k := range []string{variablesSchemaKey, outputsSchemaKey} {
+		var lsr *openapi3.SchemaRef
+
+		for _, s := range scs {
+			rsr := s[k]
+
+			// If the schema ref is not object,
+			// reset and skip.
+			if rsr == nil || rsr.Value == nil || len(rsr.Value.Properties) == 0 {
+				lsr = nil
+				break
+			}
+
+			// Remove original extension.
+			for _, p := range rsr.Value.Properties {
+				if p.Value == nil || len(p.Value.Extensions) == 0 {
+					continue
+				}
+
+				delete(p.Value.Extensions, originalExtensionKey)
+			}
+
+			// If the schema ref is nil,
+			// reuse and go ahead.
+			if lsr == nil {
+				lsr = rsr
+				continue
+			}
+
+			// Arrange the schema ref.
+			alignSchemaRef(nb, k, lsr, rsr)
+		}
+
+		if lsr != nil && len(lsr.Value.Properties) != 0 {
+			ret[k] = lsr
+		}
+	}
+
+	return ret
+}
+
+var (
+	mutuallyExclusive   struct{}
+	sentenceTerminators = sets.New('.', '!', '。', '！')
+)
+
+// alignSchemaRef aligns the schema,
+// which always move in the direction of shrinking at first, if not found, then expanding.
+func alignSchemaRef(nb map[string]any, key string, lsr, rsr *openapi3.SchemaRef) {
+	if nb == nil ||
+		lsr == nil || lsr.Value == nil ||
+		rsr == nil || rsr.Value == nil {
+		return
+	}
+
+	lv, rv := lsr.Value, rsr.Value
+
+	switch lv.Type {
+	case openapi3.TypeObject:
+		if lv.Properties != nil {
+			reqs := sets.NewString(lv.Required...)
+
+			for k := range lv.Properties {
+				// Delete the key if the type is not the same.
+				if !isSchemaRefTypeEqual(lv.Properties[k], rv.Properties[k]) {
+					delete(lv.Properties, k)
+					reqs.Delete(k)
+
+					continue
+				}
+
+				key = key + "." + k
+				if nb[key] == nil {
+					nb[key] = map[string]any{}
+				}
+
+				alignSchemaRef(nb[key].(map[string]any), key, lv.Properties[k], rv.Properties[k])
+			}
+
+			// Reset required keys.
+			if reqs.Len() == 0 {
+				lsr.Value.Required = nil
+			} else {
+				lsr.Value.Required = reqs.List()
+			}
+		} else {
+			alignSchemaRef(nb, key, lv.AdditionalProperties.Schema, rv.AdditionalProperties.Schema)
+		}
+	case openapi3.TypeArray:
+		alignSchemaRef(nb, key, lv.Items, rv.Items)
+	}
+
+	if lv.Format == "" && rv.Format != "" {
+		lv.Format = rv.Format
+	}
+
+	if !lv.WriteOnly && rv.WriteOnly {
+		lv.WriteOnly = rv.WriteOnly
+	}
+
+	// Find the most complete sentence.
+	// If not found, it cleans up the description.
```

Comment:
Can these logics be refactored into smaller functions? like alignDescription, alignDefault, alignEnum..

Reply:
although we can do this, I wouldn't say I like a lot of little pieces that function everywhere.

the independent functions are usually used for reusing, but alignDescription/alignDefault/... only works in one place.